File: api/src/main/java/rocks/milspecsg/msreferral/api/member/repository/MemberRepository.java
package rocks.milspecsg.msreferral.api.member.repository;

import rocks.milspecsg.msreferral.model.core.member.Member;
import rocks.milspecsg.msrepository.api.cache.CacheService;
import rocks.milspecsg.msrepository.api.repository.Repository;

import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public interface MemberRepository<
    TKey,
    TDataStore> extends Repository<TKey, Member<TKey>, CacheService<TKey, Member<TKey>, TDataStore>, TDataStore> {

    CompletableFuture<Optional<Member<TKey>>> getOneOrGenerateIfNotReferredForUser(UUID userUUID);

    CompletableFuture<Optional<Member<TKey>>> getOneOrGenerateForUser(UUID userUUID);

    CompletableFuture<Optional<Member<TKey>>> getOneForUser(UUID userUUID);

    CompletableFuture<Boolean> addToReferredList(TKey memberId, UUID targetUserUUID);

    CompletableFuture<Boolean> parseAndAddToReferredList(Object memberId, UUID targetUserUUID);

    CompletableFuture<Boolean> setReferred(TKey memberId, boolean referred);
}

File: referral-api/src/main/kotlin/org/anvilpowered/referral/api/member/MemberRepository.kt
/*
 *   Referral - AnvilPowered
 *   Copyright (C) 2020
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.referral.api.member

import org.anvilpowered.anvil.api.datastore.Repository
import org.anvilpowered.referral.api.model.member.Member
import java.util.UUID
import java.util.concurrent.CompletableFuture

interface MemberRepository<TKey, TDataStore> : Repository<TKey, Member<TKey>, TDataStore> {

    fun getOneIfNotReferred(userUUID: UUID): CompletableFuture<Member<TKey>?>

    fun onPlayerJoin(userUUID: UUID): CompletableFuture<Member<TKey>?>

    fun getOneForUser(userUUID: UUID): CompletableFuture<Member<TKey>?>

    fun addQueuedCommands(userUUID: UUID, commands: List<String>): CompletableFuture<Boolean>

    fun clearQueuedCommands(userUUID: UUID): CompletableFuture<Boolean>

    fun refer(userUUID: UUID, referrerUserUUID: UUID): CompletableFuture<Boolean>

    fun top(): CompletableFuture<Iterable<Member<TKey>>>
}
